File: python_various_utils/image_to_ascii.py
"""
Image to ASCII Art Converter

Converts raster images (PNG, JPEG, WEBP, BMP, GIF, TIFF) into ASCII art text files.
The core algorithm maps each pixel's grayscale luminance to a character from a
density-ordered ramp, producing a plain .txt file viewable in any monospaced font
editor or terminal.

Pipeline (step by step):

    image.png → [Load & Grayscale] → [Resize] → [Map pixels to chars] → image.txt

    1. Load & Grayscale: Open the image and convert to single-channel brightness (0-255).
    2. Resize: Shrink to target character width, applying aspect ratio correction
       (terminal chars are ~2x taller than wide, so height is scaled by ~0.55).
    3. Map pixels to chars: Each pixel brightness maps to a character in the ramp.
       Dense characters (@, #, %) represent dark pixels, light characters (., space)
       represent bright pixels.
    4. Write output: Join character rows with newlines and save as .txt file.

This module provides:

- Conversion of images to ASCII art .txt files
- Configurable output width (characters per line)
- Two character ramps: standard (10 levels) and extended (70 levels)
- Aspect ratio correction for terminal display
- Inverted mode for dark terminal backgrounds
- Single-file and recursive directory processing modes

Design Philosophy:
- Functional core with explicit helper functions
- Separate loading/resizing from character mapping
- Configurable via constants at the top

Dependencies:
- Pillow
"""

from pathlib import Path
from PIL import Image

# =========================
# CONFIG
# =========================

INPUT_FORMATS = ["png", "jpg", "jpeg", "webp", "bmp", "gif", "tiff", "tif"]

# Character ramp from darkest (most dense) to lightest (least dense)
CHAR_RAMP_STANDARD = "@%#*+=-:. "

# Extended ramp with more granularity (70 levels)
CHAR_RAMP_EXTENDED = "$@B%8&WM#*oahkbdpqwmZO0QLCJUYXzcvunxrjft/\\|()1{}[]?-_+~<>i!lI;:,\"^`'. "

# Default output width in characters
DEFAULT_WIDTH = 120

# Aspect ratio correction (terminal chars are ~2x taller than wide)
ASPECT_RATIO_CORRECTION = 0.55


# =========================
# IMAGE LOADING
# =========================

def load_and_prepare_image(path: Path) -> Image.Image:
    """
    Load an image file and convert to grayscale ('L' mode).

    Supports any format Pillow can open (PNG, JPEG, WEBP, BMP, GIF, TIFF).
    """
    img = Image.open(path)
    return img.convert("L")


# =========================
# RESIZE
# =========================

def resize_for_terminal(
    img: Image.Image,
    width: int,
    aspect_ratio_correction: float
) -> Image.Image:
    """
    Resize image to fit target character width with height correction.

    Terminal characters are taller than they are wide, so the height is
    scaled down by the correction factor to avoid vertical stretching.
    """
    original_width, original_height = img.size
    scale_factor = width / original_width
    new_height = int(original_height * scale_factor * aspect_ratio_correction)
    new_height = max(1, new_height)

    return img.resize((width, new_height), Image.Resampling.LANCZOS)


# =========================
# PIXEL TO CHARACTER MAPPING
# =========================

def map_pixels_to_chars(
    img: Image.Image,
    char_ramp: str,
    invert: bool
) -> list[str]:
    """
    Map each pixel row to a string of ASCII characters.

    Each pixel value [0, 255] is mapped to an index in the character ramp.
    In normal mode, dark pixels map to dense characters.
    In inverted mode, dark pixels map to light characters (for dark backgrounds).
    """
    pixels = list(img.getdata())
    width = img.width
    ramp_len = len(char_ramp)
    lines = []

    for row_start in range(0, len(pixels), width):
        row_pixels = pixels[row_start:row_start + width]
        row_chars = []

        for pixel_value in row_pixels:
            if invert:
                pixel_value = 255 - pixel_value

            index = pixel_value * (ramp_len - 1) // 255
            row_chars.append(char_ramp[index])

        lines.append("".join(row_chars))

    return lines


# =========================
# OUTPUT WRITER
# =========================

def write_output(lines: list[str], output_path: Path) -> None:
    """Write ASCII art lines to a text file (UTF-8)."""
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text("\n".join(lines), encoding="utf-8")


# =========================
# CONVERSION CORE
# =========================

def convert_to_ascii(
    image_path: Path,
    output_path: Path = None,
    width: int = DEFAULT_WIDTH,
    char_ramp: str = CHAR_RAMP_STANDARD,
    aspect_ratio_correction: float = ASPECT_RATIO_CORRECTION,
    invert: bool = False
) -> Path:
    """
    Full pipeline: load → grayscale → resize → map → write.

    If output_path is None, writes to the same directory with .txt extension.
    Returns the path to the output file.
    """
    if width < 1:
        raise ValueError("Width must be a positive integer.")

    if not char_ramp:
        raise ValueError("Character ramp must be a non-empty string.")

    if output_path is None:
        output_path = image_path.with_suffix(".txt")

    img = load_and_prepare_image(image_path)
    img = resize_for_terminal(img, width, aspect_ratio_correction)
    lines = map_pixels_to_chars(img, char_ramp, invert)
    write_output(lines, output_path)

    return output_path


# =========================
# SINGLE FILE MODE
# =========================

def process_single_file(
    file: str,
    width: int = DEFAULT_WIDTH,
    char_ramp: str = CHAR_RAMP_STANDARD,
    aspect_ratio_correction: float = ASPECT_RATIO_CORRECTION,
    invert: bool = False
) -> None:
    """
    Process one image file and convert to ASCII art .txt.

    Validates file exists and has a supported extension.
    """
    path = Path(file)

    if not path.exists():
        raise FileNotFoundError(f"File not found: {path}")

    ext = path.suffix.lower().replace(".", "")
    if ext not in INPUT_FORMATS:
        raise ValueError(f"Unsupported format: .{ext}. Supported: {INPUT_FORMATS}")

    try:
        out = convert_to_ascii(path, width=width, char_ramp=char_ramp,
                               aspect_ratio_correction=aspect_ratio_correction, invert=invert)
        print(f"Converted: {path} -> {out}")
    except Exception as e:
        print(f"Failed: {path} -> {e}")


# =========================
# DIRECTORY MODE
# =========================

def process_directory(
    root: str,
    source_formats: list[str] = None,
    width: int = DEFAULT_WIDTH,
    char_ramp: str = CHAR_RAMP_STANDARD,
    aspect_ratio_correction: float = ASPECT_RATIO_CORRECTION,
    invert: bool = False
) -> None:
    """
    Recursively scan a directory and convert all matching images to ASCII .txt files.

    Only files matching source_formats are processed.
    Failed conversions print an error but don't halt processing.
    """
    if source_formats is None:
        source_formats = INPUT_FORMATS

    root_path = Path(root)

    if not root_path.is_dir():
        raise NotADirectoryError(f"Not a directory: {root_path}")

    count = 0
    for file in root_path.rglob("*"):
        if file.suffix.lower().replace(".", "") in source_formats:
            try:
                out = convert_to_ascii(file, width=width, char_ramp=char_ramp,
                                       aspect_ratio_correction=aspect_ratio_correction, invert=invert)
                print(f"Converted: {file} -> {out}")
                count += 1
            except Exception as e:
                print(f"Failed: {file} -> {e}")

    print(f"\nProcessed {count} file(s).")


# =========================
# ENTRYPOINT
# =========================

if __name__ == "__main__":

    # =========================
    # CONFIG (USER-FACING)
    # =========================

    folder_to_crawl = None      # e.g. "my_images"
    single_file = "logo (16).png"          # e.g. "photo.png"

    output_width = 120          # characters per line
    use_extended_ramp = False   # True for 70-level detail, False for 10-level
    invert_brightness = False   # True for dark terminal backgrounds

    # =========================
    # RUN
    # =========================

    ramp = CHAR_RAMP_EXTENDED if use_extended_ramp else CHAR_RAMP_STANDARD

    if single_file:
        process_single_file(
            file=single_file,
            width=output_width,
            char_ramp=ramp,
            invert=invert_brightness
        )

    if folder_to_crawl:
        process_directory(
            root=folder_to_crawl,
            width=output_width,
            char_ramp=ramp,
            invert=invert_brightness
        )

    if not single_file and not folder_to_crawl:
        print("No input configured. Set 'single_file' or 'folder_to_crawl' in the CONFIG section.")

    print("\nDone.")

File: cs_various_utils/ImageToAscii.cs
// NOTE: These helper files are intended to be copied into other projects.
// Change the namespace `cs_various_utils` below to match your project's namespace
// before integrating.
//
// Image to ASCII Art Converter
//
// Converts raster images (PNG, JPEG, BMP, GIF, WEBP, TIFF) into ASCII art text files.
// The core algorithm maps each pixel's grayscale luminance to a character from a
// density-ordered ramp, producing a plain .txt file viewable in any monospaced font
// editor or terminal.
//
// Pipeline (step by step):
//
//   image.png -> [Load & Grayscale] -> [Compute Width] -> [Resize] -> [Map pixels to chars] -> image.txt
//
//   1. Load & Grayscale: Open the image and convert to single-channel brightness (0-255).
//   2. Compute Width: Read the image's native pixel width and apply the resolution
//      percentage (1-500). 100 = image width in chars, 50 = half, 200 = double.
//   3. Resize: Shrink/expand to computed character width, applying aspect ratio
//      correction (terminal chars are ~2x taller than wide, so height is scaled by ~0.55).
//   4. Map pixels to chars: Each pixel brightness maps to a character in the ramp.
//      Dense characters (@, #, %) represent dark pixels, light characters (., space)
//      represent bright pixels. The ramp is auto-selected based on effective width
//      (extended 70-level ramp for width >= 150, standard 10-level otherwise).
//   5. Write output: Join character rows with newlines and save as .txt file.
//
// This module provides:
// - Conversion of images to ASCII art .txt files
// - Resolution-based width: automatically derives output width from image dimensions
// - Resolution percentage (1-500): controls detail level relative to the source image
// - Auto ramp selection: extended (70 levels) for high-res, standard (10 levels) for low-res
// - Aspect ratio correction for terminal display
// - Inverted mode for dark terminal backgrounds
// - Single-file and recursive directory processing modes
//
// Required NuGet packages:
// - SkiaSharp
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using SkiaSharp;

// Change namespace based on whatever project is implemented
namespace cs_various_utils
{
    public static class ImageToAscii
    {
        // =========================
        // CONFIG
        // =========================

        public static readonly string[] InputFormats =
            { "png", "jpg", "jpeg", "webp", "bmp", "gif", "tiff", "tif" };

        // Character ramp from darkest (most dense) to lightest (least dense)
        public const string CharRampStandard = "@%#*+=-:. ";

        // Extended ramp with more granularity (70 levels)
        public const string CharRampExtended =
            "$@B%8&WM#*oahkbdpqwmZO0QLCJUYXzcvunxrjft/\\|()1{}[]?-_+~<>i!lI;:,\"^`'. ";

        // Default resolution percentage (100 = image's native pixel width in characters)
        public const int DefaultResolution = 100;

        // Min/max resolution bounds
        public const int MinResolution = 1;
        public const int MaxResolution = 500;

        // Threshold: use extended ramp when effective width >= this value
        public const int AutoRampThreshold = 150;

        // Aspect ratio correction (terminal chars are ~2x taller than wide)
        public const double AspectRatioCorrection = 0.55;

        // =========================
        // IMAGE LOADING
        // =========================

        /// <summary>
        /// Load an image file and convert to grayscale.
        /// Returns pixel brightness values as a 2D conceptual array via SKBitmap.
        /// </summary>
        private static SKBitmap LoadAndPrepareImage(string path)
        {
            var bitmap = SKBitmap.Decode(path)
                ?? throw new InvalidOperationException($"Failed to load image: {path}");

            // Convert to grayscale by drawing onto a Gray8 bitmap
            var gray = new SKBitmap(bitmap.Width, bitmap.Height, SKColorType.Gray8, SKAlphaType.Opaque);
            using var canvas = new SKCanvas(gray);
            using var paint = new SKPaint();
            paint.ColorFilter = SKColorFilter.CreateColorMatrix(new float[]
            {
                0.2126f, 0.7152f, 0.0722f, 0, 0,
                0.2126f, 0.7152f, 0.0722f, 0, 0,
                0.2126f, 0.7152f, 0.0722f, 0, 0,
                0,       0,       0,       1, 0
            });
            canvas.DrawBitmap(bitmap, 0, 0, paint);
            bitmap.Dispose();

            return gray;
        }

        // =========================
        // RESOLUTION LOGIC
        // =========================

        /// <summary>
        /// Compute the output character width from the image's native width
        /// and a resolution percentage (1-500).
        /// </summary>
        public static int ComputeWidthFromResolution(int imageWidth, int resolution)
        {
            resolution = Math.Max(MinResolution, Math.Min(MaxResolution, resolution));
            int width = (int)(imageWidth * (resolution / 100.0));
            return Math.Max(1, width);
        }

        /// <summary>
        /// Auto-select character ramp based on effective output width.
        /// Uses extended 70-level ramp for width >= 150, standard 10-level otherwise.
        /// </summary>
        public static string SelectCharRamp(int effectiveWidth)
        {
            return effectiveWidth >= AutoRampThreshold ? CharRampExtended : CharRampStandard;
        }

        // =========================
        // RESIZE
        // =========================

        /// <summary>
        /// Resize image to fit target character width with height correction.
        /// Terminal characters are taller than they are wide, so the height is
        /// scaled down by the correction factor to avoid vertical stretching.
        /// </summary>
        private static SKBitmap ResizeForTerminal(SKBitmap img, int width, double aspectRatioCorrection)
        {
            int originalWidth = img.Width;
            int originalHeight = img.Height;
            double scaleFactor = (double)width / originalWidth;
            int newHeight = Math.Max(1, (int)(originalHeight * scaleFactor * aspectRatioCorrection));

            var resized = img.Resize(new SKImageInfo(width, newHeight, img.ColorType, img.AlphaType),
                new SKSamplingOptions(SKCubicResampler.Mitchell));

            return resized ?? throw new InvalidOperationException("Failed to resize image.");
        }

        // =========================
        // PIXEL TO CHARACTER MAPPING
        // =========================

        /// <summary>
        /// Map each pixel row to a string of ASCII characters.
        /// Each pixel value [0, 255] is mapped to an index in the character ramp.
        /// In normal mode, dark pixels map to dense characters.
        /// In inverted mode, dark pixels map to light characters (for dark backgrounds).
        /// </summary>
        private static string[] MapPixelsToChars(SKBitmap img, string charRamp, bool invert)
        {
            int width = img.Width;
            int height = img.Height;
            int rampLen = charRamp.Length;
            var lines = new string[height];

            for (int y = 0; y < height; y++)
            {
                var row = new StringBuilder(width);
                for (int x = 0; x < width; x++)
                {
                    // Gray8 format: the color's Red channel holds the grayscale value
                    var color = img.GetPixel(x, y);
                    int pixelValue = color.Red;

                    if (invert)
                        pixelValue = 255 - pixelValue;

                    int index = pixelValue * (rampLen - 1) / 255;
                    row.Append(charRamp[index]);
                }
                lines[y] = row.ToString();
            }

            return lines;
        }

        // =========================
        // OUTPUT WRITER
        // =========================

        /// <summary>
        /// Write ASCII art lines to a text file (UTF-8).
        /// </summary>
        private static void WriteOutput(string[] lines, string outputPath)
        {
            var dir = Path.GetDirectoryName(outputPath);
            if (!string.IsNullOrEmpty(dir))
                Directory.CreateDirectory(dir);

            File.WriteAllLines(outputPath, lines, Encoding.UTF8);
        }

        // =========================
        // CONVERSION CORE
        // =========================

        /// <summary>
        /// Full pipeline: load -> grayscale -> compute width -> resize -> map -> write.
        ///
        /// Width priority:
        ///   - If outputWidth > 0: use that exact character width.
        ///   - If outputWidth &lt;= 0 (or null): derive from image width * (resolution / 100).
        ///
        /// Ramp priority:
        ///   - If useExtendedRamp is true: force extended ramp (70 levels).
        ///   - If useExtendedRamp is false: force standard ramp (10 levels).
        ///   - If useExtendedRamp is null: auto-select based on effective width.
        ///   - If charRamp is explicitly provided (non-null), it overrides everything above.
        /// </summary>
        public static string ConvertToAscii(
            string imagePath,
            string? outputPath = null,
            int? outputWidth = null,
            int resolution = DefaultResolution,
            string? charRamp = null,
            bool? useExtendedRamp = null,
            double aspectRatioCorrection = AspectRatioCorrection,
            bool invert = false)
        {
            resolution = Math.Max(MinResolution, Math.Min(MaxResolution, resolution));

            if (string.IsNullOrEmpty(outputPath))
                outputPath = Path.ChangeExtension(imagePath, ".txt");

            using var img = LoadAndPrepareImage(imagePath);

            // Determine effective width
            int effectiveWidth;
            if (outputWidth.HasValue && outputWidth.Value > 0)
            {
                effectiveWidth = outputWidth.Value;
            }
            else
            {
                effectiveWidth = ComputeWidthFromResolution(img.Width, resolution);
            }

            // Determine character ramp
            if (charRamp == null)
            {
                if (useExtendedRamp == true)
                    charRamp = CharRampExtended;
                else if (useExtendedRamp == false)
                    charRamp = CharRampStandard;
                else
                    charRamp = SelectCharRamp(effectiveWidth);
            }

            if (string.IsNullOrEmpty(charRamp))
                throw new ArgumentException("Character ramp must be a non-empty string.");

            using var resized = ResizeForTerminal(img, effectiveWidth, aspectRatioCorrection);
            var lines = MapPixelsToChars(resized, charRamp, invert);
            WriteOutput(lines, outputPath);

            var rampLabel = charRamp.Length > 10 ? "extended" : "standard";
            var widthSource = (outputWidth.HasValue && outputWidth.Value > 0) ? "manual" : $"{resolution}%";
            Console.WriteLine($"  Width: {effectiveWidth} chars ({widthSource}) | Ramp: {rampLabel}");

            return outputPath;
        }

        // =========================
        // SINGLE FILE MODE
        // =========================

        /// <summary>
        /// Process one image file and convert to ASCII art .txt.
        /// Validates file exists and has a supported extension.
        /// If outputWidth is null/0, derives width from image size * resolution%.
        /// </summary>
        public static void ProcessSingleFile(
            string file,
            int? outputWidth = null,
            int resolution = DefaultResolution,
            bool? useExtendedRamp = null,
            bool invert = false)
        {
            if (!File.Exists(file))
            {
                Console.WriteLine($"File not found: {file}");
                return;
            }

            var ext = Path.GetExtension(file).TrimStart('.').ToLowerInvariant();
            if (!InputFormats.Contains(ext))
            {
                Console.WriteLine($"Unsupported format: .{ext}. Supported: {string.Join(", ", InputFormats)}");
                return;
            }

            try
            {
                var outPath = ConvertToAscii(file, outputWidth: outputWidth,
                    resolution: resolution, useExtendedRamp: useExtendedRamp, invert: invert);
                Console.WriteLine($"Converted: {file} -> {outPath}");
            }
            catch (Exception e)
            {
                Console.WriteLine($"Failed: {file} -> {e.Message}");
            }
        }

        // =========================
        // DIRECTORY MODE
        // =========================

        /// <summary>
        /// Recursively scan a directory and convert all matching images to ASCII .txt files.
        /// If outputWidth is null/0, each image's width is derived from its own
        /// native size * resolution%.
        /// Failed conversions print an error but don't halt processing.
        /// </summary>
        public static void ProcessDirectory(
            string root,
            IEnumerable<string>? sourceFormats = null,
            int? outputWidth = null,
            int resolution = DefaultResolution,
            bool? useExtendedRamp = null,
            bool invert = false)
        {
            var formats = new HashSet<string>(
                (sourceFormats ?? InputFormats).Select(f => f.ToLowerInvariant())
            );

            if (!Directory.Exists(root))
            {
                Console.WriteLine($"Not a directory: {root}");
                return;
            }

            int count = 0;
            foreach (var file in Directory.EnumerateFiles(root, "*", SearchOption.AllDirectories))
            {
                var ext = Path.GetExtension(file).TrimStart('.').ToLowerInvariant();
                if (!formats.Contains(ext))
                    continue;

                try
                {
                    var outPath = ConvertToAscii(file, outputWidth: outputWidth,
                        resolution: resolution, useExtendedRamp: useExtendedRamp, invert: invert);
                    Console.WriteLine($"Converted: {file} -> {outPath}");
                    count++;
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Failed: {file} -> {e.Message}");
                }
            }

            Console.WriteLine($"\nProcessed {count} file(s).");
        }

        // =========================
        // ENTRYPOINT
        // =========================

        // Main execution entry point - can be adapted to command-line arguments
        // or hardcoded values for testing.
        public static void Main(string[] args)
        {
            // =========================
            // CONFIG (USER-FACING)
            // =========================

            string? folderToCrawl = null;        // e.g. "my_images"
            string? singleFile = null;           // e.g. "photo.png"

            int? outputWidth = null;             // explicit char width (overrides resolution)
                                                 // set to null to auto-derive from image size
            int resolution = 100;               // percentage of image width (1-500)
                                                 // only used when outputWidth is null
                                                 // 100 = native, 50 = half, 200 = double
            bool? useExtendedRamp = null;        // true = force extended (70 levels)
                                                 // false = force standard (10 levels)
                                                 // null = auto-select based on width
            bool invertBrightness = false;       // true for dark terminal backgrounds

            // =========================
            // RUN
            // =========================

            if (!string.IsNullOrEmpty(singleFile))
            {
                ProcessSingleFile(
                    file: singleFile,
                    outputWidth: outputWidth,
                    resolution: resolution,
                    useExtendedRamp: useExtendedRamp,
                    invert: invertBrightness
                );
            }

            if (!string.IsNullOrEmpty(folderToCrawl))
            {
                ProcessDirectory(
                    root: folderToCrawl,
                    outputWidth: outputWidth,
                    resolution: resolution,
                    useExtendedRamp: useExtendedRamp,
                    invert: invertBrightness
                );
            }

            if (string.IsNullOrEmpty(singleFile) && string.IsNullOrEmpty(folderToCrawl))
            {
                Console.WriteLine("No input configured. Set 'singleFile' or 'folderToCrawl' "
                    + "in the CONFIG section.");
            }

            Console.WriteLine("\nDone.");
        }
    }
}
